lib/grell/crawler_manager.rb

```diff
@@ -0,0 +1,72 @@
+module Grell
+  # Manages the state of the process crawling, does not care about individual pages but about logging,
+  # restarting and quiting the crawler correctly.
+  class CrawlerManager
+    # logger: logger to use for Grell's messages
+    # on_periodic_restart: if set, the driver will be restarted every :each visits (100 default), and execute the :do block
+    # driver_options: Any extra options for the Capybara driver
+    def initialize(logger: nil, on_periodic_restart: {}, driver: nil, **driver_options)
+      Grell.logger = logger ? logger : Logger.new(STDOUT)
+      @periodic_restart_block = on_periodic_restart[:do]
+      @periodic_restart_period = on_periodic_restart[:each] || PAGES_TO_RESTART
+      @driver = driver || CapybaraDriver.setup(driver_options)
+    end
+
+    # Restarts the PhantomJS process without modifying the state of visited and discovered pages.
+    def restart
+      Grell.logger.info "GRELL is restarting"
+      @driver.restart
+      Grell.logger.info "GRELL has restarted"
+    end
+
+    # Quits the poltergeist driver.
+    def quit
+      Grell.logger.info "GRELL is quitting the poltergeist driver"
+      @driver.quit
+    end
+
+    # PhantomJS seems to consume memory increasingly as it crawls, periodic restart allows to restart
+    # the driver, potentially calling a block.
+    def check_periodic_restart(collection)
+      if @periodic_restart_block && collection.visited_pages.size % @periodic_restart_period == 0
```

Comment:
is this 🆗  to not check for `periodic_restart_period > 0` ? if it happens to be zero it's going to raise a ZeroDivisionError.